### PR TITLE
[shopsys] Part of the application build is now contained in the build of the image

### PR DESCRIPTION
**Pull request by boris-brtan**
- phing build targets are added for ci and production targets
- session handler for framework is temporarily overriden during execution of phing targets

| Q             | A
| ------------- | ---
|Description, reason for the PR| to have faster deployment on CI and Production we can move some of the phing build targets into the Dockerfile that builds php-fpm image 
|New feature| No 
|BC breaks| No 
|Fixes issues| -
|Standards and tests pass| Yes
|Have you read and signed our [License Agreement for contributions](https://www.shopsys.com/license-agreement)?| Yes

### CODE REVIEW

**Comment by boris-brtan** [comment]
increase of php-fpm image for about 40MB in compressed form.
if something fails on standards or unit tests nothing will be pushed into docker registry and deploy will be now faster since no standards are executed.

**Comment by Miroslav-Stopka** [comment]
php-fpm image contains web folder also? If it does not contains web folder (for example because webserver container), then I see this modification kind of problematic

**Comment by Miroslav-Stopka** [comment]
Info for tester,

you need to test speed id CI process and Production process is faster after this modification. If the speed-up is not visible and trustworthy, this modification is propably not valuable.

File: project-base/build-dev.xml [hunk 1]
```diff
@@ -5,20 +5,21 @@
...
 
     <target name="checks" depends="db-check-mapping,standards,tests" description="Checks ORM mapping, coding standards and runs unit, DB and smoke tests."/>
     <target name="checks-diff" depends="db-check-mapping,standards-diff,tests" description="Checks ORM mapping, coding standards on changed files and runs unit, DB and smoke tests."/>
-    <target name="checks-ci" depends="db-check-mapping,standards,tests,tests-acceptance" description="Checks ORM mapping, coding standards and runs unit, DB, smoke and acceptance tests."/>
-    <target name="checks-ci-diff" depends="db-check-mapping,standards-diff,tests,tests-acceptance" description="Checks ORM mapping, coding standards on changed files and runs unit, DB, smoke and acceptance tests."/>
+    <target name="checks-ci" depends="db-check-mapping,test-db-demo,tests-functional,tests-smoke,tests-acceptance" description="Checks ORM mapping and runs DB, smoke and acceptance tests."/>
+    <target name="checks-ci-diff" depends="db-check-mapping,test-db-demo,tests-functional,tests-smoke,tests-acceptance" description="Checks ORM mapping and runs DB, smoke and acceptance tests."/>
```
[the remark this answers is not in the document]
Comment (boris-brtan) [reply project-base/build-dev.xml]: opposite, functional tests will be started but static tests will not since built image has 100% accuracy that these static tests are OK

File: project-base/build-dev.xml [hunk 2]
```diff
@@ -5,20 +5,21 @@
...
+    <target name="build-demo-ci-diff" depends="wipe-excluding-logs,timezones-check,microservices-check,dirs-create,domains-urls-check,assets,db-demo,microservice-product-search-recreate-structure,microservice-product-search-export-products,grunt,error-pages-generate,tests-acceptance-build,checks-ci-diff" description="Builds application for development with clean demo DB and runs CI checks on changed files."/>
 
     <target name="checks" depends="db-check-mapping,standards,tests" description="Checks ORM mapping, coding standards and runs unit, DB and smoke tests."/>
     <target name="checks-diff" depends="db-check-mapping,standards-diff,tests" description="Checks ORM mapping, coding standards on changed files and runs unit, DB and smoke tests."/>
-    <target name="checks-ci" depends="db-check-mapping,standards,tests,tests-acceptance" description="Checks ORM mapping, coding standards and runs unit, DB, smoke and acceptance tests."/>
-    <target name="checks-ci-diff" depends="db-check-mapping,standards-diff,tests,tests-acceptance" description="Checks ORM mapping, coding standards on changed files and runs unit, DB, smoke and acceptance tests."/>
```
[the remark this answers is not in the document]
Comment (boris-brtan) [reply project-base/build-dev.xml]: built image has now 100% accuracy that standards are OK in it so there is no need fot that now.

File: project-base/build-dev.xml [hunk 3]
```diff
@@ -5,20 +5,21 @@
...
 
     <target name="standards" depends="phplint,ecs,phpstan,twig-lint,yaml-lint,eslint-check" description="Checks coding standards."/>
     <target name="standards-diff" depends="phplint-diff,ecs-diff,phpstan,twig-lint-diff,yaml-lint,eslint-check-diff" description="Checks coding standards on changed files."/>
 
     <target name="test-db-demo" depends="clean,test-db-wipe-public-schema,test-db-import-basic-structure,test-db-migrations,test-create-domains-db-functions,test-db-fixtures-demo-singledomain,test-create-domains-data,test-generate-friendly-urls,test-db-fixtures-demo-multidomain,test-load-plugin-demo-data,test-replace-domains-urls" description="Drops all data in test database and creates a new one with demo data."/>
     <target name="test-db-performance" depends="test-db-demo,test-db-fixtures-performance" description="Drops all data in test database and creates a new one with performance data."/>
-    <target name="tests" depends="test-db-demo,tests-unit,tests-functional,tests-smoke" description="Runs unit, functional and smoke tests on a newly built test database."/>
+    <target name="tests-static" depends="tests-unit" description="Runs unit tests."/>
```
[the remark this answers is not in the document]
Comment (boris-brtan) [reply project-base/build-dev.xml]: this is for the purpose of use in monorepo and project-base it is overwritten in monorepo

File: project-base/docker/php-fpm/Dockerfile [hunk 4]
```diff
@@ -81,23 +81,33 @@ RUN composer global require hirak/prestissimo
...
+ARG project_root=.
+
 COPY --chown=www-data:www-data / /var/www/html
 
 RUN touch /var/www/html/PRODUCTION
 
 RUN composer install --optimize-autoloader --no-interaction --no-progress --no-dev
 
+# temporarily override session storage since there is no connection to redis
+RUN php phing composer npm dirs-create domains-urls-check assets
```
[the remark this answers is not in the document]
Comment (boris-brtan) [reply project-base/docker/php-fpm/Dockerfile]: now the part-1-db-independent build is not needed for production

File: project-base/docker/php-fpm/Dockerfile [hunk 5]
```diff
@@ -81,23 +81,33 @@ RUN composer global require hirak/prestissimo
...
+
 ########################################################################################################################
 
 FROM base as ci
 
+ARG project_root=.
```
[the remark this answers is not in the document]
Comment (boris-brtan) [reply project-base/docker/php-fpm/Dockerfile]: removed

File: project-base/build-dev.xml [hunk 6]
```diff
@@ -5,20 +5,21 @@
...
     <target name="build-demo-dev-quick" depends="wipe-excluding-logs,composer-dev,npm,dirs-create,domains-urls-check,assets,db-demo,microservice-product-search-recreate-structure,microservice-product-search-export-products,grunt,tests-acceptance-build" description="Builds application for development with clean demo DB while skipping nonessential steps."/>
-    <target name="build-demo-ci" depends="wipe-excluding-logs,composer-dev,npm,timezones-check,microservices-check,dirs-create,domains-urls-check,assets,db-demo,microservice-product-search-recreate-structure,microservice-product-search-export-products,grunt,error-pages-generate,tests-acceptance-build,checks-ci" description="Builds application for development with clean demo DB and runs CI checks."/>
-    <target name="build-demo-ci-diff" depends="wipe-excluding-logs,composer-dev,npm,timezones-check,microservices-check,dirs-create,domains-urls-check,assets,db-demo,microservice-product-search-recreate-structure,microservice-product-search-export-products,grunt,error-pages-generate,tests-acceptance-build,checks-ci-diff" description="Builds application for development with clean demo DB and runs CI checks on changed files."/>
```
[the remark this answers is not in the document]
Comment (boris-brtan) [reply project-base/build-dev.xml]: missed that one, removed.
for CI everything is copied from source-code into var/www/html and on production there is created named volume from web folder

File: project-base/docker/php-fpm/Dockerfile [hunk 7]
```diff
@@ -81,23 +81,33 @@ RUN composer global require hirak/prestissimo
...
 # see project-base/kubernetes/deployments/webserver-php-fpm.yml for details
 COPY --chown=www-data:www-data / /source-code
 WORKDIR /source-code
 RUN composer install --optimize-autoloader --no-interaction --no-progress
 
+# clean cache because of recreating valid autloading of vendor after /source-code is copied into /var/www/html
+RUN php phing composer-dev npm dirs-create domains-urls-check assets tests-static clean
+
 # switch back to the original workdir
 WORKDIR /var/www/html
```
[the remark this answers is not in the document]
Comment (boris-brtan) [reply project-base/docker/php-fpm/Dockerfile]: i think that no.

File: project-base/docker/php-fpm/Dockerfile (hunk 4 above)
[the remark this answers is not in the document]
Comment (boris-brtan) [reply project-base/docker/php-fpm/Dockerfile]: we have discussion with @TomasLudvik and we came to the decistion that is is OK

File: project-base/docker/php-fpm/Dockerfile [hunk 8]
```diff
@@ -87,6 +87,9 @@ RUN touch /var/www/html/PRODUCTION
 
 RUN composer install --optimize-autoloader --no-interaction --no-progress --no-dev
 
+# temporarily override session storage since there is no connection to redis
```
Comment (TomasLudvik) [inline project-base/docker/php-fpm/Dockerfile]: This comment does not make sense.

**Comment by TomasLudvik** [comment]
I have tested this configuration on my local with target production and this is output:

```
90678fbe1bce_shopsys-framework-php-fpm | mkfifo: /tmp/log-pipe: File exists
90678fbe1bce_shopsys-framework-php-fpm exited with code 1
```

File: project-base/docker/php-fpm/docker-php-entrypoint [hunk 9]
```diff
@@ -6,11 +6,11 @@ set -e
 
 # reading from a named pipe and writing to stdout via "tail -f"
 # this workaround prevents logs from being output to console during console commands
+# multistage build uses different UIDs and logpipe is transformed into something else than pipe so best way is to recreate it
 PIPE=/tmp/log-pipe
-if ! [ -p $PIPE ]; then
-    mkfifo $PIPE
-    chmod 660 $PIPE
-fi
+rm -rf $PIPE
```
Comment (TomasLudvik) [inline project-base/docker/php-fpm/docker-php-entrypoint]: I think this should be in separate commit describing why is this done

**Comment by boris-brtan** [comment]
rebased on master because of new build-dev.xml configs

File: UPGRADE.md [hunk 10]
```diff
@@ -113,6 +113,7 @@ There is a list of all the repositories maintained by monorepo, changes in log b
     - rebuild image by running `docker-compose up -d --build`
     - files `build.xml` and `build-dev.xml` were updated to speed up deployment process of built docker images of php-fpm
     - installation guide for production via Docker was updated, now there is no need for the first part of the build phing target
+    - file `docker-php-entrypoint` was changed, update it from [`project-base/docker/php-fpm/docker-php-entrypoint`](./project-base/docker/php-fpm/docker-php-entrypoint)
```
Comment (Miroslav-Stopka) [inline UPGRADE.md]: change `docker-php-entrypoint` to full path

update it from => according to
